include/zephyr/drivers/charger.h

```diff
@@ -0,0 +1,221 @@
+	/** Value should be of type bool*/
+	CHARGER_PROP_PRESENT,
+	/** Represents the charging status of the charger. */
+	/** Value should be of type enum charger_status */
+	CHARGER_PROP_STATUS,
+	/** Reserved to demark end of common charger properties */
+	CHARGER_PROP_COMMON_COUNT,
+	/**
+	 * Reserved to demark downstream custom properties - use this value as the actual value may
+	 * change over future versions of this API
+	 */
+	CHARGER_PROP_CUSTOM_BEGIN = CHARGER_PROP_COMMON_COUNT + 1,
+	/** Reserved to demark end of valid enum properties */
+	CHARGER_PROP_MAX = UINT16_MAX,
+};
+
+/**
+ * @typedef charger_prop_t
+ * @brief A charger property's identifier
+ *
+ * See charger_property for a list of identifiers
+ */
+typedef uint16_t charger_prop_t;
+
+/**
+ * @brief External supply states
+ */
+enum charger_online {
+	/** External supply not present */
+	CHARGER_ONLINE_OFFLINE = 0,
+	/** External supply is present and of fixed output */
+	CHARGER_ONLINE_FIXED,
+	/** External supply is present and of programmable output*/
+	CHARGER_ONLINE_PROGRAMMABLE,
+};
+
+/**
+ * @brief Charging states
+ */
+enum charger_status {
+	/** Charging device state is unknown */
+	CHARGER_STATUS_UNKNOWN = 0,
+	/** Charging device is charging a battery */
+	CHARGER_STATUS_CHARGING,
+	/** Charging device is not able to charge a battery */
+	CHARGER_STATUS_DISCHARGING,
+	/** Charging device is not charging a battery */
+	CHARGER_STATUS_NOT_CHARGING,
+	/** The battery is full and the charging device will not attempt charging */
+	CHARGER_STATUS_FULL,
+};
+
+struct charger_get_property {
+	/** Battery charger property to get */
+	charger_prop_t property_type;
+
+	/** Negative error status set by callee e.g. -ENOTSUP for an unsupported property */
+	int ret;
+
+	/** Property field for getting */
+	union {
+		/* Fields have the format: */
+		/* CHARGER_PROPERTY_FIELD */
+		/* type property_field; */
+
+		/* Dynamic Battery Info */
+		/** CHARGER_PROP_ONLINE */
+		enum charger_online online;
+		/** CHARGER_PROP_PRESENT */
+		bool present;
+		/** CHARGER_PROP_STATUS */
+		enum charger_status status;
+	} value;
+};
+
+struct charger_set_property {
+	/** Battery charger property to set */
+	charger_prop_t property_type;
+
+	/** Negative error status set by callee e.g. -ENOTSUP for an unsupported property */
+	int ret;
+
+	/** Property field for getting */
+	union {
+		/* Fields have the format: */
+		/* CHARGER_PROPERTY_FIELD */
+		/* type property_field; */
+
+		/* Dynamic Battery Info */
+		/** CHARGER_PROP_ONLINE */
+		enum charger_online online;
+		/** CHARGER_PROP_PRESENT */
+		bool present;
+		/** CHARGER_PROP_STATUS */
+		enum charger_status status;
+	} value;
+};
+
+/**
+ * @typedef charger_get_property_t
+ * @brief Callback API for getting a charger property.
+ *
+ * See charger_get_property() for argument description
+ */
+typedef int (*charger_get_property_t)(const struct device *dev, struct charger_get_property *props,
```

Comment:
Let's just make this use a single property struct. Per https://github.com/zephyrproject-rtos/zephyr/pull/56666#discussion_r1301472409 discussion.

If someone eventually wants a multi-property it won't be hard to add it in the future.

Reply:
ack